### ISSUE TITLE
Recent java.util.NoSuchElementException problems with Account.getMessages(...)

### ISSUE THREAD

**Issue by cgbaker** [twilio/twilio-java#204]
I've been struggling with a problem. I'm using the Twilio Java SDK (multiple version, from 3.6.1 to 4.2.0) via Scala, in a Play Framework application. I've been checking for messages to a particular Twilio number via the following:
```
val messageList: MessageList = owningAccount.getMessages(Map("To" -> number))
```

Recently, this has started resulting in NoSuchElementException if I try to do anything with the returned list (e.g., .size, .foreach). It didn't occur until recently (it was working on Friday, April 8, was not working when I checked again today). I was working under the assumption that something in my build environment had somehow changed, but I've tested now with multiple versions of Scala 2.11.x and multiple versions of Java (1.6, 1.7 and 1.8), and I'm still getting the issue.

For the above, the number is "+19739754967", and upon returning from getMessages(), messageList.getNumPages() returns 0 (thought this was strange; I know there is a message), while messageList.size or any attempt to use the iterator (directly Java-style or via Scala's JavaConversions) results in the NoSuchElementException.

I tried switching my code to the following:
```
val messageList: Iterable[instance.Message] = owningAccount.getMessages().filter(_.getTo == number)
```
This is obviously less efficient, doing the filtering on my end and not benefit from Twilio's presumably more sophisticated matching; but it allowed me to loop over all my messages, which I had previously been unable to do.

I wish I could provide a full example to demonstrate this, but I'm on a bit of a time crunch. I'll stick with my current approach for now, but if I'm doing something wrong, please let me know.

**Comment by skimbrel** [comment]
Hey Chris,

Sorry you've been running into this. Can you try making the same request against the REST API with `curl` or the like and seeing what comes back (in particular the list metadata attributes)? The first step here is always to figure out whether there's an issue with the API proper or just with the library.

As a heads-up, `getNumPages()` should have been removed from the library with the 4.0 release (see https://www.twilio.com/engineering/2015/04/16/replacing-absolute-paging-with-relative-paging for the background here) but it looks like we missed that one in the cleanup effort.

**Comment by cgbaker** [comment]
Sure thing. 

Without a request parameter for To, "next_page_uri" is null, "num_pages" is 1, and "total" is 16 (I don't understand this, because there's only two messages returned.)

With a request parameter for To specifying my number, "num_pages" is 2, "next_page_uri" has a URI referring to Page=1, and "total" is 64 (another number that I don't understand.) If I hit the "next_page_uri" referenced here, I get a payload whose metadata is consistent, but which contains no messages. Which makes a certain amount of sense; there were only two messages returned, and they appeared among the "first 50" on Page=1.

So, the Java SDK seems fine. It makes the reasonable assumption that, after parsing NextPage, there will be messages present (there aren't, hence the failure on getNext()). 

As for the discrepancy between "total": 16 when there's only 2, without a "To" filter; and "total": 64 when there's only 2 with a "To" filter; it seems to be the root. This also suggests that as I progress and the "total" for my filter-less call goes past 50, I'll start seeing the same problem. It also suggests another workaround; just pass a large "pagesize" parameter.

Another detail: this account is a sub-account to which the number is assigned. Not sure whether that's relevant. I assume that Twilio support can look up my account info via the number posted above.

Thanks for the debugging tip.

**Comment by skimbrel** [comment]
Hmm, I have a better idea now. Next question — have you deleted any message records for this account? I think we noted this in the blog post I linked, but one of the subtle bugs we're trying to address with removing counts is that the cached list counts can't be updated when records get deleted (a new-ish feature).

**Comment by cgbaker** [comment]
Yes. I delete messages after I see them.

**Comment by skimbrel** [comment]
All right. That would cause this behavior — the counts are stale (because we don't have the ability to guess which sets of filters have been used against the collection to go and invalidate the cached counts), so the API is generating a `next_page_uri` in the response when it shouldn't be, causing the Java library to think it has more elements available when it doesn't. I can *probably* concoct a fix for this in the library, but the right behavior is for the API to produce proper metadata (i.e. without a `next_page_uri`) in this case.

Again, really sorry you're running into this — I'm as frustrated as you that the API isn't behaving itself properly here, and we'll get a fix out for you ASAP.

**Comment by cgbaker** [comment]
I appreciate it. I'll use a larger page count for now to avoid hitting this. We're not production on this yet, so it's not the end of this world.

**Comment by cgbaker** [comment]
Any update?

**Comment by cgbaker** [comment]
Okay, I've been maxing out the PageSize query parameter, which is limited by the API to 1000. Unfortunately, I've hit a limit where I'm seeing the bug. I don't know how long it is before the caches are refreshed, but I feel like I'm seeing it on rarely trafficked accounts now. The only further workaround that I can identify is to anticipate it in my code and trap the exception. 

We're looking at a product release in the September timeline, so any estimate on a resolution would be appreciated. (I understand how hard it is to make such estimates, I'm asking in spite of that ;)

Thank you.

**Comment by georgepalli** [comment]
+1,

I'm also running into this when deleting messages. I am using the Twilio Java SDK. I iterate over MessageList and called message.delete(). At some point in the iteration I get the java.util.NoSuchElementException 

Here is the stacktrace
`
[java.util.ArrayList$Itr.next(ArrayList.java:854),
com.twilio.sdk.resource.ListResource$ListIterator.next(ListResource.java:60),
com.twilio.sdk.resource.ListResource$ListIterator.next(ListResource.java:21),
com.example.package.TwilioCleanupManager ...
`

**Comment by jingming** [comment]
@cgbaker this should be addressed in release 5.7.1! I'm trying this with: 
`
TwilioRestClient client = new TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN);

Map<String, String> filters = new HashMap<String, String>();
filters.put("PageSize", "1");

CallList calls = client.getAccount().getCalls(filters);
for (Call call : calls) {
    System.out.println(call.getSid());
}
`
and it outputs:
```
CA2c4a4fee583204dd134378677c17425c
CA8db1ab523c9141a4e25f8d88e26f0fcd
...
CA255f98d362c9ec87740926ce5b732931

Process finished with exit code 0
```

@georgepalli
for the purposes of mass deletion, we suggest fetching the first page, deleting all the entries, and then refetching the first page until you are out. 
`
MessageList messages = client.getAccount().getMessages();
List<Message> pageData = messages.getPageData();

while(!pageData.isEmpty()) {
    for (Message message : pageData) {
        message.delete();
    }

    messages = client.getAccount().getMessages();
    pageData = messages.getPageData();
}
`

**Comment by jingming** [comment]
Answered.
